### PR TITLE
Optimize i32(x) % C_pot in boolean context

### PR DESCRIPTION
**Pull request by MaxGraey**
`bool(i32(x) % C_pot)` -> `bool(i32(x) & (C_pot - 1))`
`bool(i32(x) % min_s)` -> `bool(i32(x) & max_s)`

For all other situations we already do this for `(i32|i64).rem_s`

### CODE REVIEW

**Comment by MaxGraey** [comment]
Fuzzed:
```
Invocations so far:
   FuzzExec: 11762
   CompareVMs: 2079
   CheckDeterminism: 1008
   Wasm2JS: 1213
   Asyncify: 1961

ITERATION: 13028
```

**Comment by MaxGraey** [comment]
@tlively could you take a look at this? Thanks)

File: src/passes/OptimizeInstructions.cpp [hunk 1]
```diff
@@ -994,6 +994,20 @@ struct OptimizeInstructions
           //   return binary;
           // }
         }
+      } else if (binary->op == RemSInt32) {
+        // bool((signed)x % C_pot)  ==>  bool(x & (C_pot - 1))
```
Comment (kripken) [inline src/passes/OptimizeInstructions.cpp:998]: The case of `isSignedMin` is not documented in the comment here?

About that case of signed min, isn't it true that
```
(signed)x % -2147483648  ==>  x == -2147483648
```
even outside a boolean context?

Comment (MaxGraey) [reply src/passes/OptimizeInstructions.cpp:998]: > About that case of signed min, isn't it true that
> ```
> (signed)x % -2147483648  ==>  x == -2147483648
> ```

No, let's see how LLVM handle this:
https://godbolt.org/z/K51Knj

So inside boolean context it `(signed)x % min` -> `(signed)x & max`

Comment (kripken) [reply src/passes/OptimizeInstructions.cpp:998]: Oh, right, I was thinking of division...

Comment (MaxGraey) [reply src/passes/OptimizeInstructions.cpp:1056]: Ah, sorry. Forget about this. Done!

File: test/passes/optimize-instructions_all-features.wast [hunk 2]
```diff
@@ -4517,6 +4517,15 @@
...
+    ;; (signed)x % 4 ? 1 : 0
+    (drop (if (result i32)
+      (i32.rem_s
+        (local.get $x)
+        (i32.const 4)
+      )
+      (i32.const 1)
+      (i32.const 0)
+    ))
   )
```
Comment (kripken) [inline test/passes/optimize-instructions_all-features.wast:4529]: oh, also, please add a testcase here for the min/max.

lgtm with that, sorry for not noticing before.

**Review by kripken — APPROVED** [review]

File: test/passes/optimize-instructions_all-features.wast (hunk 2 above)
Comment (MaxGraey) [reply test/passes/optimize-instructions_all-features.wast:4538]: Done!

Comment (kripken) [reply test/passes/optimize-instructions_all-features.wast:4538]: Great!

**Pull request merged**